pkg/storage/oss/oss.go

```diff
@@ -0,0 +1,671 @@
+		err := svc.SetBucketACL(d.Config.Bucket, oss.ACLPrivate)
+
+		if err != nil {
+			if oerr, ok := err.(oss.ServiceError); ok {
+				util.UpdateCondition(cr, defaults.StoragePublicAccessBlocked, operatorapi.ConditionFalse, oerr.Code, oerr.Error())
+			} else {
+				util.UpdateCondition(cr, defaults.StoragePublicAccessBlocked, operatorapi.ConditionFalse, "Unknown Error Occurred", err.Error())
+			}
+		} else {
+			util.UpdateCondition(cr, defaults.StoragePublicAccessBlocked, operatorapi.ConditionTrue, "Public Access Block Successful", "Public access to the OSS bucket and its contents have been successfully blocked.")
+			cr.Status.Storage = imageregistryv1.ImageRegistryConfigStorage{
+				OSS: d.Config.DeepCopy(),
+			}
+			cr.Spec.Storage.OSS = d.Config.DeepCopy()
+		}
+	}
+
+	// Tag the bucket with the openshiftClusterID
+	// along with any user defined tags from the cluster configuration
+	if cr.Spec.Storage.ManagementState == imageregistryv1.StorageManagementStateManaged {
+		klog.Info("setting Alibaba Cloud OSS bucket tags")
+
+		tagset := []oss.Tag{
+			{
+				Key:   "kubernetes.io/cluster/" + infra.Status.InfrastructureName,
+				Value: "owned",
+			},
+			{
+				Key:   "Name",
+				Value: infra.Status.InfrastructureName + "-image-registry",
+			},
+			{
+				Key:   "sigs.k8s.io/cloud-provider-alibaba/origin",
+				Value: "ocp",
+			},
+			{
+				Key:   "GISV",
+				Value: "ocp",
+			},
+		}
+
+		// at this stage we are not keeping user tags in sync. as per enhancement proposal
+		// we only set user provided tags when we created the bucket.
+		hasOSSStatus := infra.Status.PlatformStatus != nil && infra.Status.PlatformStatus.AlibabaCloud != nil
+		if hasOSSStatus {
+			klog.Infof("user provided %d tags", len(infra.Status.PlatformStatus.AlibabaCloud.ResourceTags))
+			for _, tag := range infra.Status.PlatformStatus.AlibabaCloud.ResourceTags {
+				klog.Infof("user provided bucket tag: %s: %s", tag.Key, tag.Value)
+				tagset = append(tagset, oss.Tag{
+					Key:   tag.Key,
+					Value: tag.Value,
+				})
+			}
+		}
+		klog.V(5).Infof("tagging bucket with tags: %+v", tagset)
+
+		tagging := oss.Tagging{
+			Tags: tagset,
+		}
+		err := svc.SetBucketTagging(d.Config.Bucket, tagging)
+		if err != nil {
+			if oerr, ok := err.(oss.ServiceError); ok {
+				util.UpdateCondition(cr, defaults.StorageTagged, operatorapi.ConditionFalse, oerr.Code, oerr.Error())
+			} else {
+				util.UpdateCondition(cr, defaults.StorageTagged, operatorapi.ConditionFalse, "Unknown Error Occurred", err.Error())
+			}
+		} else {
+			util.UpdateCondition(cr, defaults.StorageTagged, operatorapi.ConditionTrue, "Tagging Successful", "Tags were successfully applied to the OSS bucket")
+		}
+	} else {
+		klog.Info("ignoring bucket tags, storage is not managed")
+	}
+
+	// Enable default encryption on the bucket
+	if cr.Spec.Storage.ManagementState == imageregistryv1.StorageManagementStateManaged {
```

Comment:
@menglingwei I was reviewing the s3 code from here https://github.com/openshift/cluster-image-registry-operator/blob/master/pkg/storage/s3/s3.go#L673-L720.

I think we need to take into account the KMS encryption. We can check the encryption type from https://github.com/openshift/api/pull/1082/files#diff-c5fb014eef97ced7b6ba12508db516e2454667ce34820b705992420cded4d723R338 and then perform the necessary actions to set the correct type of encryption along with the KMS keyid (if KMS is used).

Reply:
@kwoodson  I think support KMS is essential.